src/nvim/diff.c

```diff
@@ -2062,28 +2092,216 @@ static void run_linematch_algorithm(diff_T *dp)
       // we add it to the array of char*, diffbufs
       diffbufs[ndiffs] = diffbufs_mm[ndiffs].ptr;
 
-      // keep track of the length of this diff block to pass it to the linematch
-      // algorithm
-      diff_length[ndiffs] = dp->df_count[i];
+      diff_lines[ndiffs] = dp->df_count[i];
+      if (diff_allignment == CHARMATCH || diff_allignment == WORDMATCH) {
+        // before removing whitespace for charmatch
+        result_diff_start_pos[ndiffs] = total_chars_length;
+        // get the length of each of the diffs
+        int lines = dp->df_count[i];
+        const char *p = diffbufs[ndiffs];
+        while (lines) {
+          total_chars_length++;  // increment the total characters counter
+          if (*p == '\n') {
+            lines--;
+          }
+          p++;
+        }
+      } else if (diff_allignment == LINEMATCH) {
+        // LINEMATCH
+        // keep track of the length of this diff block to pass it to the linematch
+        // algorithm
+        diff_length[ndiffs] = dp->df_count[i];
+      }
 
       // increment the amount of diff buffers we are passing to the algorithm
       ndiffs++;
     }
   }
 
+  // allocate all the memory we will need to keep track of tokens positions and their respective
+  // lengths. For word matching, this is the 'word' as vim defines it, for character matching, the
+  // token is the one or more 8 bit 'chars' that make up a utf character
+  if (diff_allignment == WORDMATCH || diff_allignment == CHARMATCH) {
+    // are we ignoring whitespace in the comparison?
+    if (iwhite) {
+      // allocate array for index mapping of result array
+      iwhite_index_offset = xmalloc(total_chars_length * sizeof(size_t));
+    }
+    for (size_t i = 0; i < ndiffs; i++) {
+      word_offset[i] = xmalloc(total_chars_length * sizeof(size_t));
+      word_offset_size[i] = xmalloc(total_chars_length * sizeof(size_t));
+      for (size_t j = 0; j < total_chars_length; j++) {
+        word_offset_size[i][j] = 0;
+      }
+    }
+  }
+  // calculate the token lengths and white space offset and pre process the contents of the diffs to
+  // remove white space if necessary
+  for (size_t i = 0; i < ndiffs; i++) {
```

Comment:
this big block of processing produces a result. probably makes sense to extract it to a function?